### ISSUE TITLE
OpenSSL cert loading performance issue

### ISSUE THREAD

**Issue by cyberw** [geventhttpclient/geventhttpclient#219]
Python 3.12 / OpenSSL 3.0 has issues on windows, which was reported for Locust: https://github.com/locustio/locust/issues/2555

It isnt specific to geventhttpclient, but may require a workaround on our side, as suggested here for requests:

https://github.com/psf/requests/pull/6667

**Comment by ml31415** [comment]
I'm not sure if this applies to us. If I read it correctly, the SSL context creation is the thing that can take some time. But we're already creating the context on a per-pool basis and not per-connection.
https://github.com/geventhttpclient/geventhttpclient/blob/50a9482edda5216929ac2ad09c1e7f9062372b2c/geventhttpclient/connectionpool.py#L248

I don't know how the mentioned "50 peak users" are constructed. But if all of them share one UserAgent and therefore one HTTPClient and its connection pool, there should be no problem. If every user gets its own UserAgent, well than you might have to fix that for Locust.

**Comment by cyberw** [comment]
Well... We kind of want it to be realistic (one ssl handshake, one http connections per locust User). Here's a slimmed down example:

**Comment by cyberw** [comment]
The above code takes 0.2s on Python 3.11 and takes around 7s on 3.12 (on windows). It isnt geventhttpclient's fault of course, but perhaps it can be optimized somehow...

**Comment by ml31415** [comment]
Yeah, but again: geventhttpclient is doing its best to reduce overhead by having only one context per host per session. The most reasonable thing, that comes to mind would be some kind of shallow copy of a UserAgent. One which copies an agent including the hosts of a client pool and their context, but without the open sockets. Would be a rather weird thing to have, and you might just have that in Locust instead of in geventhttpclient.

**Comment by cyberw** [comment]
I can try implementing that. Thanks. Ok to close if you want.
